Added a `Sink` implementation to `stream::FromErr`.

46bit opened the pull request:
This support is very useful from my experience using the combinator with `Sink + Stream`.

It [simplifies](https://github.com/sirpent-team/comms/blob/master/src/client/mod.rs#L47-L49) a type from `
Client<I, Unsplit<SinkFromErr<SplitSink<C>, ErrorString>, FromErr<SplitStream<C>, ErrorString>>>` to `Client<I, FromErr<C>>`. Without this patch that'll be the general case of using it with, e.g., `TcpStream` - not pleasant at all.

I'm not sure if you want this in the library, so feel free to decline if you'd rather.

alexcrichton commented:
Thanks for the PR! A number of other stream combinators also forward `Sink` implementations but they do so transparently. Perhaps this could transparently forward the `Sink` implementation (like other stream combinators) and then this could be achieved with `stream.from_err().sink_from_err()` perhaps?

46bit commented:
> forward Sink implementations

Aha, that's the terminology I was looking for.

I agree that would be a better implementation; I'll get it done :-)

Cheers,
Michael

On 24 March 2017 at 15:40, Alex Crichton <notifications@github.com> wrote:

> Thanks for the PR! A number of other stream combinators also forward Sink
> implementations but they do so transparently. Perhaps this could
> transparently forward the Sink implementation (like other stream
> combinators) and then this could be achieved with
> stream.from_err().sink_from_err() perhaps?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexcrichton/futures-rs/pull/426#issuecomment-289058783>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJ9J8jk7krpKjL_tBWtMOnjEDmgtW_6ks5ro-PrgaJpZM4Mnf8y>
> .
>

alexcrichton commented:
I'm curious, did you agree that we should split this into `sink_from_err`? Or do you feel like it's best to just use this one combinator?

46bit commented:
Hey, sorry, this has been on my back burner. I definitely think the two combinators is the best way forward - larger type signatures, but much more flexible usage. I'll update this before go to sleep.

46bit commented:
I've updated this so that `stream::FromErr` and `sink::SinkFromErr preserve `Sink + Stream`, and added basic tests. There's no test for preserving `Sink + Stream` as I wasn't sure where in the testsuite it belongs.

alexcrichton commented:
Thanks!
